src/System.Net.Http/src/System/Net/Http/StreamToStreamCopy.cs

```diff
@@ -2,59 +2,266 @@
+            Debug.Assert(destination != null);
+            Debug.Assert(bufferSize > 0);
 
-            byte[] buffer = new byte[bufferSize];
+            try
+            {
+                // If the source is a MemoryStream from which we can extract the internal buffer,
+                // then we can perform the copy in a single write operation.
+                ArraySegment<byte> sourceBuffer;
+                MemoryStream sourceMemoryStream = source as MemoryStream;
+                if (sourceMemoryStream != null && sourceMemoryStream.TryGetBuffer(out sourceBuffer))
+                {
+                    // It's possible source derives from MemoryStream but has a bad override of Position.
+                    // If we get back a value that doesn't make sense, don't take the optimized path.
+                    long pos = source.CanSeek ? source.Position : -1;
+                    if (pos >= 0 && pos < sourceBuffer.Count)
+                    {
+                        // We need to copy from the current position, so if necessary update the buffer
+                        // based on the position of the stream.
+                        if (pos != 0)
+                        {
+                            sourceBuffer = new ArraySegment<byte>(
+                                sourceBuffer.Array, 
+                                (int)checked(sourceBuffer.Offset + pos),
+                                (int)checked(sourceBuffer.Count - pos));
+                        }
+
+                        // Update position to simulate reading all the data
+                        source.Position += sourceBuffer.Count; 
+
+                        // Now write the buffer to the destination stream. This will complete synchronously if the 
+                        // destination's WriteAsync completes synchronously, such as if destination is also a MemoryStream.  
+                        // Thus we don't need to special case it.
+                        return WriteToAnyStreamAsync(sourceBuffer, source, destination, disposeSource);
+                    }
+                }
+
+                // The source is not a MemoryStream whose buffer we can use directly, but the destination might be our special 
+                // LimitMemoryStream, in which case if it's been pre-sized with a capacity, we can optimize the copy 
+                // by giving its buffer to the source directly, avoiding the need for an intermediate buffer.
+                HttpContent.LimitMemoryStream destinationLimitStream = destination as HttpContent.LimitMemoryStream;
+                if (destinationLimitStream != null)
+                {
+                    int capacity = destinationLimitStream.Capacity;
+                    if (capacity > 0 && destinationLimitStream.MaxSize >= capacity)
+                    {
+                        return CopyAsyncToPreSizedLimitMemoryStream(source, destinationLimitStream, bufferSize, disposeSource);
+                    }
```

Comment:
I don't understand the if here.  The capacity > 0 I can understand; but why does a relationship between MaxSize and Capacity exist?  Length, perhaps... but Capacity?